stan/math/prim/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,68 @@
+#ifndef STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/chol2inv.hpp>
+#include <stan/math/prim/fun/add_diag.hpp>
+#include <stan/math/prim/fun/rep_vector.hpp>
+#include <stan/math/prim/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/multiply_lower_tri_self_transpose.hpp>
+#include <stan/math/prim/fun/mdivide_left_tri_low.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the Moore-Penrose generalized inverse of the specified matrix.
+ *
+ * @tparam T type of elements in the matrix
+ * @tparam n number of rows, can be Eigen::Dynamic
+ * @tparam m number of columns, can be Eigen::Dynamic
+ *
+ * @param M specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ */
+template <typename EigMat,
+          require_eigen_vt<std::is_arithmetic, EigMat>* = nullptr>
```

Comment:
I do plan to add the rev implementation once I understand how to do it :). Going to work with @SteveBronder on it.

Reply:
This should work for rev even without an actual specialization so I think you can make this just `require_eigen_t<EigMat>`. Once we have the rev specialization we can add a `require_not_vt_var<EigMat>` (that checks that the `value_type` is not var)

Reply:
Now that the rev part is in, do I just add `require_not_vt_var` in prim instead of `require_eigen_t`? Also, for the alias that inputs a scalar is there anything to do there (ie `require_stan_scalar_t<Scal>`)?

Reply:
@t4c1 @SteveBronder I spent some time trying to get the tests to work with `require_not_vt_var` but failed. The tests using `require_eigen_t` passed on my end with what I pushed this morning.